wgpu-core/src/device/queue.rs

```diff
@@ -1133,6 +1133,12 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                 .fetch_add(1, Ordering::Relaxed)
                 + 1;
             let mut active_executions = Vec::new();
+
+            // SAFETY: We're constructing this during the submission phase,
+            // where all resources it uses are guaranteed to outlive this
+            // short-lived set.
+            let mut submit_surface_textures = A::SubmitSurfaceTextureSet::new();
```

Comment:
I don't understand why insert needs to be unsafe, it's just a vector

Reply:
> Insertion is unsafe because it can own raw resources related to surfaces (e.g. the wait semaphore).

I believe this is actually safe. The backends aren't acutally storing references, they're copying in handles which do nothing on drop.

The unsafe-ness comes from the `use` of the raw resources - in this case the queue.submit in which it is passed. Similarly this is why `ash::*Builder::build()` is safe even though it creates a bunch of raw pointers, it's the _use_ of those pointers which is unsafe, and the functions it is passed to are unsafe.

More generally, I think we should just pass a normal slice into `queue.submit` - removing the collection of surface textures seems not really worth the wgpu-hal complexity.

Reply:
> More generally, I think we should just pass a normal slice into queue.submit - removing the collection of surface textures seems not really worth the wgpu-hal complexity.

Resource tracking has changed quite a bit since your branch, this requires breaking up `pending_writes` (since we would borrow from it and it currently would need mutable access to [`pre_submit` here](https://github.com/gfx-rs/wgpu/pull/4967/files#diff-150156a37cf3627465ceb22096ed995ee26ae640007c421e726134bafd499dbeR1492)) and dealing with how texture handles are handed it here in `cmd_buf_trackers`:

```
error[E0597]: `texture` does not live long enough
    --> wgpu-core\src\device\queue.rs:1227:59
     |
1225 | ...                   for texture in cmd_buf_trackers.textures.used_resources() {
     |                           ------- binding `texture` declared here
1226 | ...                       let id = texture.info.id();
1227 | ...                       let should_extend = match texture.inner.get(&snatch_guard) {
     |                                                     ^^^^^^^ borrowed value does not live long enough
...
1240 | ...                                   submit_surface_textures.push(raw);
     |                                       --------------------------------- borrow later used here
...
1261 | ...                   }
     |                       - `texture` dropped here while still borrowed
```
Holding onto `used_resources` in a separate collection would get this down. Alternatively use the unsafe APIs related to accessing tracked textures.

Finally this means that every backend needs to do this work, even though they don't need to access their surface textures.

Does this sound about what you expect and do you want me to move forward with the change?

Reply:
Hmmmmm

My resistance to this is solely because having a container trait in wgpu-hal feels very wrong - we should be able to do what we need without needing that.

There are potential solutions where we clone the texture, and extract the surface texture _after_ pre_submit, maybe?

As it stands, this is not obviously sound - it happens to be sound because the snatch guard is held open, but it would be nice to have a raw resource attached by lifetime to the actual snatch guard, so it can't be messed up.

I'm not too worried about the cost of doing it "unnecessarily" for other backends, we do that a lot.

Reply:
So this might be a workable solution, we can change `Snatchable<T>` so that `pub fn get()` works like this instead (introducing the `'snatch`) lifetime:

```rust
pub fn get<'snatch>(&self, _guard: &'snatch SnatchGuard) -> Option<&'snatch T> {
    unsafe { (*self.value.get()).as_ref() }
}
```

That would mean that references to the `A::SurfaceTexture` would outlive the loop and live as long as the snatch guard instead of the handle it was taken from, allowing them to be used. This does seem like the intent of the snatch guard.

How does this sound to you?

Reply:
I've reworked it to do this (in a separate commit), tell me what you think.

Reply:
So I'm pretty sure now that lifetime is wrong. Since Snatchable types can be destructed independently of the lock, allowing such a lifetime would allow them to be used after they've been freed.

Reply:
I've opted now to do the naive thing so we can ship this: take a clone of the texture and reborrow before passing it in.